File: frontend/src/lib/api/pim.ts
import { apiClient } from "./client";

// ============================================================================
// TypeScript Interfaces
// ============================================================================

export interface PimIgmpJoin {
  group: string;
  source_addresses: string[];
}

export interface PimInterfaceIgmp {
  disabled: boolean;
  joins: PimIgmpJoin[];
  query_interval: number | null;
  query_max_response_time: number | null;
  version: number | null;
}

export interface PimInterface {
  name: string;
  bfd: boolean;
  bfd_profile: string | null;
  dr_priority: number | null;
  hello: number | null;
  no_bsm: boolean;
  no_unicast_bsm: boolean;
  passive: boolean;
  source_address: string | null;
  igmp: PimInterfaceIgmp | null;
}

export interface PimRpAddress {
  address: string;
  groups: string[];
}

export interface PimRp {
  addresses: PimRpAddress[];
  keep_alive_timer: number | null;
}

export interface PimSptSwitchover {
  infinity_and_beyond: boolean;
  prefix_list: string | null;
}

export interface PimConfig {
  ecmp: boolean;
  ecmp_rebalance: boolean;
  igmp_watermark_warning: number | null;
  interfaces: PimInterface[];
  join_prune_interval: number | null;
  keep_alive_timer: number | null;
  no_v6_secondary: boolean;
  packets: number | null;
  register_accept_list_prefix_list: string | null;
  register_suppress_time: number | null;
  rp: PimRp | null;
  spt_switchover: PimSptSwitchover | null;
  ssm_prefix_list: string | null;
}

export interface PimCapabilities {
  version: string;
  features: Record<string, { supported: boolean; description: string }>;
  version_info: {
    is_1_4: boolean;
    is_1_5: boolean;
  };
  instance_name?: string;
  instance_id?: string;
}

export interface PimBatchOperation {
  op: string;
  value?: string;
}

export interface PimBatchGroup {
  interface?: string;
  operations: PimBatchOperation[];
}

export interface PimBatchRequest {
  groups: PimBatchGroup[];
}

export interface VyOSResponse {
  success: boolean;
  data?: Record<string, unknown> | null;
  error?: string | null;
}

// ============================================================================
// Global Settings Shape (for inline edit diffing)
// ============================================================================

export interface PimGlobalSettings {
  ecmp: boolean;
  ecmp_rebalance: boolean;
  igmp_watermark_warning: string;
  join_prune_interval: string;
  keep_alive_timer: string;
  no_v6_secondary: boolean;
  packets: string;
  register_accept_list_prefix_list: string;
  register_suppress_time: string;
  spt_infinity_and_beyond: boolean;
  spt_prefix_list: string;
  ssm_prefix_list: string;
}

// ============================================================================
// API Service
// ============================================================================

class PimService {
  async getCapabilities(): Promise<PimCapabilities> {
    return apiClient.get<PimCapabilities>("/vyos/pim/capabilities");
  }

  async getConfig(refresh = false): Promise<PimConfig> {
    return apiClient.get<PimConfig>("/vyos/pim/config", {
      refresh: refresh.toString(),
    });
  }

  async refreshConfig(): Promise<VyOSResponse> {
    return apiClient.post<VyOSResponse>("/vyos/config/refresh");
  }

  async batchConfigure(request: PimBatchRequest): Promise<VyOSResponse> {
    const result = await apiClient.post<VyOSResponse>("/vyos/pim/batch", request);
    if (!result.success) {
      throw new Error(result.error || "Operation failed");
    }
    await this.refreshConfig();
    return result;
  }

  // ==========================================================================
  // Global Settings
  // ==========================================================================

  async updateGlobalSettings(
    current: PimGlobalSettings,
    next: PimGlobalSettings
  ): Promise<VyOSResponse> {
    const ops: PimBatchOperation[] = [];

    // ECMP
    if (next.ecmp !== current.ecmp) {
      ops.push({ op: next.ecmp ? "set_ecmp" : "delete_ecmp" });
    }
    // ECMP rebalance
    if (next.ecmp_rebalance !== current.ecmp_rebalance) {
      ops.push({ op: next.ecmp_rebalance ? "set_ecmp_rebalance" : "delete_ecmp_rebalance" });
    }
    // If ECMP turned off, also delete rebalance
    if (!next.ecmp && current.ecmp_rebalance) {
      ops.push({ op: "delete_ecmp_rebalance" });
    }

    // no_v6_secondary
    if (next.no_v6_secondary !== current.no_v6_secondary) {
      ops.push({ op: next.no_v6_secondary ? "set_no_v6_secondary" : "delete_no_v6_secondary" });
    }

    // Timers (string fields)
    this.diffStringField(ops, current.join_prune_interval, next.join_prune_interval, "set_join_prune_interval", "delete_join_prune_interval");
    this.diffStringField(ops, current.keep_alive_timer, next.keep_alive_timer, "set_keep_alive_timer", "delete_keep_alive_timer");
    this.diffStringField(ops, current.register_suppress_time, next.register_suppress_time, "set_register_suppress_time", "delete_register_suppress_time");
    this.diffStringField(ops, current.packets, next.packets, "set_packets", "delete_packets");
    this.diffStringField(ops, current.igmp_watermark_warning, next.igmp_watermark_warning, "set_igmp_watermark_warning", "delete_igmp_watermark_warning");

    // Prefix lists
    this.diffStringField(ops, current.register_accept_list_prefix_list, next.register_accept_list_prefix_list, "set_register_accept_list_prefix_list", "delete_register_accept_list_prefix_list");
    this.diffStringField(ops, current.ssm_prefix_list, next.ssm_prefix_list, "set_ssm_prefix_list", "delete_ssm_prefix_list");

    // SPT switchover
    if (next.spt_infinity_and_beyond !== current.spt_infinity_and_beyond) {
      if (next.spt_infinity_and_beyond) {
        ops.push({ op: "set_spt_switchover_infinity" });
      } else {
        ops.push({ op: "delete_spt_switchover" });
      }
    }
    if (next.spt_infinity_and_beyond) {
      this.diffStringField(ops, current.spt_prefix_list, next.spt_prefix_list, "set_spt_switchover_infinity_prefix_list", "delete_spt_switchover_infinity_prefix_list");
    } else if (current.spt_prefix_list) {
      // Clearing prefix list when infinity unchecked is handled by delete_spt_switchover above
    }

    if (ops.length === 0) {
      return { success: true, data: { message: "No changes" } };
    }

    return this.batchConfigure({ groups: [{ operations: ops }] });
  }

  private diffStringField(
    ops: PimBatchOperation[],
    currentVal: string,
    nextVal: string,
    setOp: string,
    deleteOp: string
  ) {
    if (nextVal !== currentVal) {
      if (nextVal.trim()) {
        ops.push({ op: setOp, value: nextVal.trim() });
      } else if (currentVal.trim()) {
        ops.push({ op: deleteOp });
      }
    }
  }

  // ==========================================================================
  // Interface Operations
  // ==========================================================================

  async createInterface(iface: PimInterface): Promise<VyOSResponse> {
    const ops = this.buildInterfaceOps(iface);
    return this.batchConfigure({
      groups: [{ interface: iface.name, operations: ops }],
    });
  }

  async updateInterface(existing: PimInterface, updated: PimInterface): Promise<VyOSResponse> {
    // Delete-and-recreate is simpler for nested IGMP joins
    const deleteOps: PimBatchOperation[] = [{ op: "delete_interface" }];
    const createOps = this.buildInterfaceOps(updated);
    return this.batchConfigure({
      groups: [
        { interface: existing.name, operations: deleteOps },
        { interface: updated.name, operations: createOps },
      ],
    });
  }

  async deleteInterface(name: string): Promise<VyOSResponse> {
    return this.batchConfigure({
      groups: [{ interface: name, operations: [{ op: "delete_interface" }] }],
    });
  }

  private buildInterfaceOps(iface: PimInterface): PimBatchOperation[] {
    const ops: PimBatchOperation[] = [{ op: "set_interface" }];

    if (iface.bfd) {
      ops.push({ op: "set_interface_bfd" });
      if (iface.bfd_profile) {
        ops.push({ op: "set_interface_bfd_profile", value: iface.bfd_profile });
      }
    }
    if (iface.dr_priority != null) {
      ops.push({ op: "set_interface_dr_priority", value: String(iface.dr_priority) });
    }
    if (iface.hello != null) {
      ops.push({ op: "set_interface_hello", value: String(iface.hello) });
    }
    if (iface.no_bsm) {
      ops.push({ op: "set_interface_no_bsm" });
    }
    if (iface.no_unicast_bsm) {
      ops.push({ op: "set_interface_no_unicast_bsm" });
    }
    if (iface.passive) {
      ops.push({ op: "set_interface_passive" });
    }
    if (iface.source_address) {
      ops.push({ op: "set_interface_source_address", value: iface.source_address });
    }

    // IGMP
    if (iface.igmp) {
      ops.push({ op: "set_interface_igmp" });
      if (iface.igmp.disabled) {
        ops.push({ op: "set_interface_igmp_disable" });
      }
      if (iface.igmp.query_interval != null) {
        ops.push({ op: "set_interface_igmp_query_interval", value: String(iface.igmp.query_interval) });
      }
      if (iface.igmp.query_max_response_time != null) {
        ops.push({ op: "set_interface_igmp_query_max_response_time", value: String(iface.igmp.query_max_response_time) });
      }
      if (iface.igmp.version != null) {
        ops.push({ op: "set_interface_igmp_version", value: String(iface.igmp.version) });
      }
      for (const join of iface.igmp.joins) {
        ops.push({ op: "set_interface_igmp_join", value: join.group });
        for (const src of join.source_addresses) {
          ops.push({ op: "set_interface_igmp_join_source", value: `${join.group},${src}` });
        }
      }
    }

    return ops;
  }

  // ==========================================================================
  // RP Operations
  // ==========================================================================

  async createRpAddress(rp: PimRpAddress): Promise<VyOSResponse> {
    const ops: PimBatchOperation[] = [{ op: "set_rp_address", value: rp.address }];
    for (const group of rp.groups) {
      ops.push({ op: "set_rp_address_group", value: `${rp.address},${group}` });
    }
    return this.batchConfigure({ groups: [{ operations: ops }] });
  }

  async updateRpAddress(existing: PimRpAddress, updated: PimRpAddress): Promise<VyOSResponse> {
    const ops: PimBatchOperation[] = [];

    // Remove old groups
    for (const g of existing.groups) {
      if (!updated.groups.includes(g)) {
        ops.push({ op: "delete_rp_address_group", value: `${existing.address},${g}` });
      }
    }
    // Add new groups
    for (const g of updated.groups) {
      if (!existing.groups.includes(g)) {
        ops.push({ op: "set_rp_address_group", value: `${updated.address},${g}` });
      }
    }

    if (ops.length === 0) {
      return { success: true, data: { message: "No changes" } };
    }
    return this.batchConfigure({ groups: [{ operations: ops }] });
  }

  async deleteRpAddress(address: string): Promise<VyOSResponse> {
    return this.batchConfigure({
      groups: [{ operations: [{ op: "delete_rp_address", value: address }] }],
    });
  }

  async updateRpKeepAliveTimer(value: string, currentValue: string): Promise<VyOSResponse> {
    const ops: PimBatchOperation[] = [];
    if (value.trim() !== currentValue.trim()) {
      if (value.trim()) {
        ops.push({ op: "set_rp_keep_alive_timer", value: value.trim() });
      } else if (currentValue.trim()) {
        ops.push({ op: "delete_rp_keep_alive_timer" });
      }
    }
    if (ops.length === 0) {
      return { success: true, data: { message: "No changes" } };
    }
    return this.batchConfigure({ groups: [{ operations: ops }] });
  }
}

export const pimService = new PimService();

File: backend/routers/pim/pim.py
"""
PIM Router

API endpoints for managing VyOS PIM (Protocol Independent Multicast) configuration.
No version differences between VyOS 1.4 and 1.5.

Config tree:
  protocols pim/
    ecmp
    ecmp rebalance
    igmp/
      watermark-warning  (1-65535)
    interface/<IFACE>/
      bfd
      bfd profile        (txt)
      dr-priority        (1-4294967295)
      hello              (1-180)
      igmp/
        disable
        join/<GROUP>/
          source-address (multi, ipv4)
        query-interval           (1-1800)
        query-max-response-time  (10-250)
        version                  (2|3)
      no-bsm
      no-unicast-bsm
      passive
      source-address     (ipv4)
    join-prune-interval  (1-65535)
    keep-alive-timer     (1-65535)
    no-v6-secondary
    packets              (1-255)
    register-accept-list/
      prefix-list        (txt)
    register-suppress-time (1-65535)
    rp/
      address/<IP>/
        group            (multi, ipv4net)
      keep-alive-timer   (1-65535)
    spt-switchover/
      infinity-and-beyond
      infinity-and-beyond prefix-list (txt)
    ssm/
      prefix-list        (txt)
"""

from fastapi import APIRouter, HTTPException, Request
from starlette.concurrency import run_in_threadpool
from pydantic import BaseModel, Field
from typing import List, Dict, Optional, Any
from session_vyos_service import get_session_vyos_service
from vyos_builders.pim import PimBatchBuilder
from fastapi_permissions import require_read_permission, require_write_permission
from rbac_permissions import FeatureGroup
import inspect
import logging

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/vyos/pim", tags=["pim"])


# ============================================================================
# Pydantic Models
# ============================================================================


class PimIgmpJoin(BaseModel):
    """IGMP join group on an interface."""
    group: str
    source_addresses: List[str] = []


class PimInterfaceIgmp(BaseModel):
    """IGMP settings for a PIM interface."""
    disabled: bool = False
    joins: List[PimIgmpJoin] = []
    query_interval: Optional[int] = None
    query_max_response_time: Optional[int] = None
    version: Optional[int] = None


class PimInterface(BaseModel):
    """PIM interface configuration."""
    name: str
    bfd: bool = False
    bfd_profile: Optional[str] = None
    dr_priority: Optional[int] = None
    hello: Optional[int] = None
    no_bsm: bool = False
    no_unicast_bsm: bool = False
    passive: bool = False
    source_address: Optional[str] = None
    igmp: Optional[PimInterfaceIgmp] = None


class PimRpAddress(BaseModel):
    """Rendezvous Point address configuration."""
    address: str
    groups: List[str] = []


class PimRp(BaseModel):
    """RP configuration."""
    addresses: List[PimRpAddress] = []
    keep_alive_timer: Optional[int] = None


class PimSptSwitchover(BaseModel):
    """SPT switchover configuration."""
    infinity_and_beyond: bool = False
    prefix_list: Optional[str] = None


class PimConfig(BaseModel):
    """Complete PIM configuration."""
    ecmp: bool = False
    ecmp_rebalance: bool = False
    igmp_watermark_warning: Optional[int] = None
    interfaces: List[PimInterface] = []
    join_prune_interval: Optional[int] = None
    keep_alive_timer: Optional[int] = None
    no_v6_secondary: bool = False
    packets: Optional[int] = None
    register_accept_list_prefix_list: Optional[str] = None
    register_suppress_time: Optional[int] = None
    rp: Optional[PimRp] = None
    spt_switchover: Optional[PimSptSwitchover] = None
    ssm_prefix_list: Optional[str] = None


class PimBatchOperation(BaseModel):
    """Single operation in a batch request."""
    op: str = Field(..., description="Operation name")
    value: Optional[str] = Field(None, description="Operation value")


class PimBatchGroup(BaseModel):
    """A group of operations scoped to a single interface (or global)."""
    interface: Optional[str] = Field(None, description="Interface name (null for global ops)")
    operations: List[PimBatchOperation]


class PimBatchRequest(BaseModel):
    """Model for batch configuration. Supports multiple groups in a single atomic commit."""
    groups: List[PimBatchGroup]


class VyOSResponse(BaseModel):
    """Standard response from VyOS operations."""
    success: bool
    data: Optional[Dict[str, Any]] = None
    error: Optional[str] = None


# ============================================================================
# Endpoint 1: Capabilities
# ============================================================================


@router.get("/capabilities")
async def get_pim_capabilities(request: Request):
    """Get PIM feature capabilities based on device VyOS version."""
    await require_read_permission(request, FeatureGroup.PIM)

    try:
        service = get_session_vyos_service(request)
        version = service.get_version()
        builder = PimBatchBuilder(version=version)
        capabilities = builder.get_capabilities()

        if hasattr(request.state, "instance") and request.state.instance:
            capabilities["instance_name"] = request.state.instance.get("name")
            capabilities["instance_id"] = request.state.instance.get("id")
        return capabilities
    except Exception:
        logger.exception("Unhandled error")
        raise HTTPException(status_code=500, detail="Internal server error")


# ============================================================================
# Endpoint 2: Config
# ============================================================================


@router.get("/config", response_model=PimConfig)
async def get_pim_config(http_request: Request, refresh: bool = False):
    """Get all PIM configuration from VyOS in a generalized format."""
    await require_read_permission(http_request, FeatureGroup.PIM)

    try:
        service = get_session_vyos_service(http_request)
        full_config = await run_in_threadpool(service.get_full_config, refresh=refresh)

        pim_config = full_config.get("protocols", {}).get("pim", {})

        if not pim_config:
            return PimConfig()

        return PimConfig(
            ecmp="ecmp" in pim_config and (
                pim_config["ecmp"] is None
                or isinstance(pim_config.get("ecmp"), dict)
                or pim_config.get("ecmp") == {}
            ),
            ecmp_rebalance=_has_nested(pim_config, "ecmp", "rebalance"),
            igmp_watermark_warning=_safe_int(_deep_get(pim_config, "igmp", "watermark-warning")),
            interfaces=_parse_interfaces(pim_config.get("interface", {})),
            join_prune_interval=_safe_int(pim_config.get("join-prune-interval")),
            keep_alive_timer=_safe_int(pim_config.get("keep-alive-timer")),
            no_v6_secondary="no-v6-secondary" in pim_config,
            packets=_safe_int(pim_config.get("packets")),
            register_accept_list_prefix_list=_deep_get(pim_config, "register-accept-list", "prefix-list"),
            register_suppress_time=_safe_int(pim_config.get("register-suppress-time")),
            rp=_parse_rp(pim_config.get("rp")),
            spt_switchover=_parse_spt_switchover(pim_config.get("spt-switchover")),
            ssm_prefix_list=_deep_get(pim_config, "ssm", "prefix-list"),
        )
    except Exception:
        logger.exception("Unhandled error")
        raise HTTPException(status_code=500, detail="Internal server error")


# ============================================================================
# Config Parsers
# ============================================================================


def _deep_get(data: dict, *keys: str):
    """Safely traverse nested dicts."""
    for key in keys:
        if not isinstance(data, dict):
            return None
        data = data.get(key)
        if data is None:
            return None
    return data


def _safe_int(value) -> Optional[int]:
    """Convert value to int safely."""
    if value is None:
        return None
    try:
        return int(value)
    except (ValueError, TypeError):
        return None


def _ensure_list(value) -> List[str]:
    """Convert VyOS config value to a list."""
    if value is None:
        return []
    if isinstance(value, list):
        return value
    if isinstance(value, str):
        return [value]
    return []


def _has_nested(data: dict, *keys: str) -> bool:
    """Check if nested key path exists (for flag-style nodes like ecmp.rebalance)."""
    for key in keys:
        if not isinstance(data, dict):
            return False
        val = data.get(key)
        if val is None and key not in data:
            return False
        data = val if val is not None else {}
    return True


def _parse_interfaces(interfaces_raw: dict) -> List[PimInterface]:
    """Parse PIM interface configurations."""
    if not interfaces_raw:
        return []

    interfaces = []
    for iface_name, iface_config in interfaces_raw.items():
        if iface_config is None:
            iface_config = {}

        igmp_config = iface_config.get("igmp")
        igmp = None
        if igmp_config is not None:
            if not isinstance(igmp_config, dict):
                igmp_config = {}
            igmp = PimInterfaceIgmp(
                disabled="disable" in igmp_config,
                joins=_parse_igmp_joins(igmp_config.get("join", {})),
                query_interval=_safe_int(igmp_config.get("query-interval")),
                query_max_response_time=_safe_int(igmp_config.get("query-max-response-time")),
                version=_safe_int(igmp_config.get("version")),
            )

        bfd_config = iface_config.get("bfd")
        has_bfd = bfd_config is not None
        bfd_profile = None
        if isinstance(bfd_config, dict):
            bfd_profile = bfd_config.get("profile")

        interfaces.append(PimInterface(
            name=iface_name,
            bfd=has_bfd,
            bfd_profile=bfd_profile,
            dr_priority=_safe_int(iface_config.get("dr-priority")),
            hello=_safe_int(iface_config.get("hello")),
            no_bsm="no-bsm" in iface_config,
            no_unicast_bsm="no-unicast-bsm" in iface_config,
            passive="passive" in iface_config,
            source_address=iface_config.get("source-address"),
            igmp=igmp,
        ))

    return interfaces


def _parse_igmp_joins(joins_raw: dict) -> List[PimIgmpJoin]:
    """Parse IGMP join group configurations."""
    if not joins_raw or not isinstance(joins_raw, dict):
        return []

    joins = []
    for group_addr, join_config in joins_raw.items():
        if join_config is None:
            join_config = {}
        joins.append(PimIgmpJoin(
            group=group_addr,
            source_addresses=_ensure_list(join_config.get("source-address") if isinstance(join_config, dict) else None),
        ))
    return joins


def _parse_rp(rp_raw) -> Optional[PimRp]:
    """Parse RP configuration."""
    if not rp_raw or not isinstance(rp_raw, dict):
        return None

    addresses = []
    address_raw = rp_raw.get("address", {})
    if isinstance(address_raw, dict):
        for addr, addr_config in address_raw.items():
            if addr_config is None:
                addr_config = {}
            addresses.append(PimRpAddress(
                address=addr,
                groups=_ensure_list(addr_config.get("group") if isinstance(addr_config, dict) else None),
            ))

    return PimRp(
        addresses=addresses,
        keep_alive_timer=_safe_int(rp_raw.get("keep-alive-timer")),
    )


def _parse_spt_switchover(spt_raw) -> Optional[PimSptSwitchover]:
    """Parse SPT switchover configuration."""
    if not spt_raw or not isinstance(spt_raw, dict):
        return None

    infinity_config = spt_raw.get("infinity-and-beyond")
    has_infinity = infinity_config is not None or "infinity-and-beyond" in spt_raw
    prefix_list = None
    if isinstance(infinity_config, dict):
        prefix_list = infinity_config.get("prefix-list")

    return PimSptSwitchover(
        infinity_and_beyond=has_infinity,
        prefix_list=prefix_list,
    )


# ============================================================================
# Endpoint 3: Batch Operations
# ============================================================================


@router.post("/batch", response_model=VyOSResponse)
async def pim_batch_configure(http_request: Request, body: PimBatchRequest):
    """Execute a batch of PIM configuration operations.

    Accepts multiple groups, each optionally scoped to a single interface.
    All groups are processed into a single builder and committed atomically.
    """
    await require_write_permission(http_request, FeatureGroup.PIM)

    try:
        service = get_session_vyos_service(http_request)
        version = service.get_version()
        builder = PimBatchBuilder(version=version)

        for group in body.groups:
            for operation in group.operations:
                method = getattr(builder, operation.op)
                sig = inspect.signature(method)
                params = [p for p in sig.parameters.keys() if p != "self"]

                args = []

                if "iface" in params and group.interface:
                    args.append(group.interface)

                if operation.value and len(params) > len(args):
                    args.append(operation.value)

                method(*args)

        if builder.is_empty():
            return VyOSResponse(success=True, data={"message": "No operations to execute"})

        response = service.execute_batch(builder)

        return VyOSResponse(
            success=response.status == 200,
            data={"message": "PIM configuration updated"},
            error=response.error if response.error else None,
        )
    except AttributeError as e:
        raise HTTPException(status_code=400, detail=f"Unknown operation: {str(e)}")
    except Exception:
        logger.exception("Unhandled error")
        raise HTTPException(status_code=500, detail="Internal server error")
